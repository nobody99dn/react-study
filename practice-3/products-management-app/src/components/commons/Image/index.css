.image {
  height: auto;
  max-width: 100%;
}

/* - Logo */
/* - ---------------------------------------------- */
.image-logo {
  background-repeat: no-repeat;
  width: 6.25rem;
  text-indent: -9999px;
}

/* - Icon */
/* - ---------------------------------------------- */
.image-icon {
  display: block;
  background-repeat: no-repeat;
  width: 2.25rem;
  cursor: pointer;
}

.icon-left {
  position: absolute;
  right: 1.5rem;
}

.close-icon {
  position: absolute;
  cursor: pointer;
  right: 1rem;
  top: 1rem;
}

/* - Card image */
/* - ---------------------------------------------- */
.card-image {
  border-radius: 5px;
  object-fit: contain;
  overflow: hidden;
  height: 18rem;
}

.image-error {
  line-height: 1.5rem;
}
