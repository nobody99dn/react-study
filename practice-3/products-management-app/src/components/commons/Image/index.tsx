// Library
import React, { memo, useState } from 'react';

// Styles
import './index.css';

// Constants
import { ERROR_MESSAGES, ImageVariants } from '@constants/index';
import Text from '@components/commons/Text';

interface ImageProps {
  alt: string;
  className?: string;
  height?: string;
  imageUrl: string;
  width?: string;
  variant?: ImageVariants;
  isError?: boolean;
  onImageClick?: () => void;
  onImageError?: () => void;
}

const Image: React.FC<ImageProps> = ({
  alt,
  className,
  height,
  imageUrl,
  width,
  variant = ImageVariants.Default,
  isError = false,
  onImageClick,
  onImageError
}) => {
  return !isError ? (
    <img
      className={`image ${
        variant !== ImageVariants.Default ? `image-${variant} ` : ''
      } ${className}`}
      alt={alt}
      src={imageUrl}
      style={{ width: width, height: height }}
      onClick={onImageClick}
      onError={onImageError}
    />
  ) : (
    <div className='image-error'>
      <Text color='red'>{ERROR_MESSAGES.IMAGE_NOT_FOUND}</Text>
    </div>
  );
};

export default memo(Image);
