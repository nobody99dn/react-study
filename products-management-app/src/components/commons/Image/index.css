.image {
  height: auto;
  max-width: 100%;
}

.image-logo {
  background-repeat: no-repeat;
  display: block;
  width: 6.25rem;
  text-indent: -9999px;
}

.image-icon {
  background-repeat: no-repeat;
  display: block;
}

.icon-left {
  position: absolute;
  right: 1.5rem;
}

.icon-left {
  position: absolute;
  right: 1.5rem;
}

.card-image {
  border-radius: 5px;
}
