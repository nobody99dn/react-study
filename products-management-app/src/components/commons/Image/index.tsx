import React from 'react';

import './index.css';

// Constants
import { ImageVariants } from '@/constants/types';

interface ImageProps {
  alt: string;
  className?: string;
  height?: string;
  imageUrl: string;
  width?: string;
  variant?: ImageVariants;
  onImageClick?: () => void;
}

export const Image: React.FC<ImageProps> = (
  {
    alt,
    className,
    height,
    imageUrl,
    width,
    variant = ImageVariants.Default,
    onImageClick
  }
) => (
  <img
    className={
      ['image',
        variant !== ImageVariants.Default
          ? `image-${variant} `
          : '',
        className
      ].join(' ').trim()
    }
    alt={alt}
    src={imageUrl}
    style={{ width: width, height: height }}
    onClick={onImageClick}
  />
);
