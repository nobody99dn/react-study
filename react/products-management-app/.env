VITE_BASE_URL=https://62943cf0a7203b3ed0650c9b.mockapi.io
