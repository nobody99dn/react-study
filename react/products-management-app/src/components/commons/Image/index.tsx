// Library
import React, { memo } from 'react';

// Styles
import './index.css';

// Constants
import { ImageVariants } from '@constants/index';

interface ImageProps {
  alt: string;
  className?: string;
  height?: string;
  imageUrl: string;
  width?: string;
  variant?: ImageVariants;
  onImageClick?: () => void;
}

const Image: React.FC<ImageProps> = ({
  alt,
  className,
  height,
  imageUrl,
  width,
  variant = ImageVariants.Default,
  onImageClick
}) => (
  <img
    className={`image ${
      variant !== ImageVariants.Default ? `image-${variant} ` : ''
    } ${className}`}
    alt={alt}
    src={imageUrl}
    style={{ width: width, height: height }}
    onClick={onImageClick}
  />
);

export default memo(Image);
