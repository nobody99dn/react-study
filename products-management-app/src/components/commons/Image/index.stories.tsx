import { ComponentStory, ComponentMeta } from '@storybook/react';

import { Image, ImageTypes } from './index';

import '../../../assets/styles/App.css';
import '../../../assets/styles/variables.css';
import './index.css';

import logo from '../../../assets/images/logo.png';
import iPhoneImage from '../../../assets/images/iphone-13-promax-alpine-green-alpine-green-color.jpg';
import deleteIcon from '../../../assets/images/icons/trash-fill.svg';

export default {
  title: 'Components/Image',
  component: Image
} as ComponentMeta<typeof Image>;

const Template: ComponentStory<typeof Image> = (args) => <Image {...args} />;

export const Default = Template.bind({});
Default.args = {
  imageUrl: iPhoneImage,
  width: '10rem'
};

export const Logo = Template.bind({});
Logo.args = {
  imageUrl: logo,
  alt: 'Logo',
  variant: ImageTypes.Logo
};

export const Icon = Template.bind({});
Icon.args = {
  imageUrl: deleteIcon,
  alt: 'Delete icon',
  variant: ImageTypes.Icon
};
