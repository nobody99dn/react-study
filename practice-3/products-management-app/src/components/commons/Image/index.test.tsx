// Libraries
import { render, act, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';

// Components
import Image from './index';

// Image
import { blankImage } from '@assets/index';
import { ImageVariants } from '@constants/index';

describe('button component', () => {
  let container: HTMLElement;

  const defaultProps = {
    alt: 'image',
    imageUrl: blankImage
  };

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultImage = create(<Image {...defaultProps} />).toJSON();
    expect(defaultImage).toMatchSnapshot();

    const logo = create(
      <Image {...defaultProps} variant={ImageVariants.Logo} />
    ).toJSON();
    expect(logo).toMatchSnapshot();

    const icon = create(
      <Image {...defaultProps} variant={ImageVariants.Icon} />
    ).toJSON();
    expect(icon).toMatchSnapshot();
  });
});
