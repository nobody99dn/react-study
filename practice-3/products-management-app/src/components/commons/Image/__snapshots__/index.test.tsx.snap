// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component should render correctly 1`] = `
<img
  alt="image"
  className="image"
  onError={[Function]}
  onLoad={[Function]}
/>
`;

exports[`Image component should render correctly 2`] = `
<img
  alt="image"
  className="image image-logo"
  onError={[Function]}
  onLoad={[Function]}
/>
`;

exports[`Image component should render correctly 3`] = `
<img
  alt="image"
  className="image image-icon"
  onError={[Function]}
  onLoad={[Function]}
/>
`;
